### ISSUE TITLE
issue when define variables

### ISSUE THREAD

**Issue by wangzhezhe** [ornladios/ADIOS2#2076]
Hello, 

I got an issue when I define variables, this is related code

```
...
  adios2::Variable<double> var_u;
    adios2::Variable<int> var_step;

    size_t shapex = bs.m_indexub[0] - bs.m_indexlb[0] + 1;
    size_t shapey = bs.m_indexub[1] - bs.m_indexlb[1] + 1;
    size_t shapez = bs.m_indexub[2] - bs.m_indexlb[2] + 1;

    sprintf(str, "adios write step %s varName %s start %d %d %d count %d %d %d\n",
            parameters[0].c_str(), parameters[1].c_str(),
            bs.m_indexlb[0], bs.m_indexlb[1], bs.m_indexlb[2],
            shapex, shapey, shapez);
    std::cout << str << std::endl;

    //there is problem to call the defineVaraible in parallel
    fmr->m_statefulConfig->m_adiosLock.lock();
    var_u = fmr->m_statefulConfig->m_io.DefineVariable<double>("U",
                                                               {512, 512, 512},
                                                               {(size_t)bs.m_indexlb[0], (size_t)bs.m_indexlb[1], (size_t)bs.m_indexlb[2]},
                                                               {shapex, shapey, shapez});

    int step = atoi(parameters[0].c_str());
    var_step = fmr->m_statefulConfig->m_io.DefineVariable<int>("step");
    fmr->m_statefulConfig->m_adiosLock.unlock();

    fmr->m_statefulConfig->m_writer.Put<int>(var_step, &step);
    fmr->m_statefulConfig->m_writer.Put<double>(var_u, (double *)inputData);
...
```
Basically, the data is stored in-memory beforehand and when sth happens, the corresponding data is written out by adios, the function of data writing may called by multiple threads in parallel,  and I init the adios by this way:

```
void initADIOS()
        {
                adios2::ADIOS adios(adios2::DebugON);
                this->m_io = adios.DeclareIO("gorilla_gs");
                this->m_io.SetEngine("BP4");
                this->m_writer = m_io.Open("gorilla_gs", adios2::Mode::Write);
        }
```
and I got some errors like this:

```
adios write step 1 varName grascott_u start 256 0 256 count 256 256 256


Thread 4 "unimos_server" received signal SIGFPE, Arithmetic exception.
[Switching to Thread 0x2aaab9bc5700 (LWP 33051)]
0x00000000200589bb in std::__detail::_Mod_range_hashing::operator() (
    this=0x100006c8950, __num=11876154508389848687, __den=0)
    at /opt/gcc/8.3.0/snos/include/g++/bits/hashtable_policy.h:434
434	    { return __num % __den; }
Missing separate debuginfos, use: zypper install libbz2-1-debuginfo-1.0.6-5.9.1.x86_64 libelf1-debuginfo-0.168-4.5.3.x86_64 libexpat1-debuginfo-2.2.5-3.6.1.x86_64 libffi7-debuginfo-3.2.1.git259-3.16.x86_64 libltdl7-debuginfo-2.4.6-1.406.x86_64 libpng16-16-debuginfo-1.6.34-3.9.1.x86_64 libz1-debuginfo-1.2.11-3.9.1.x86_64 unixODBC-debuginfo-2.3.6-1.12.x86_64
bt
(gdb) #0  0x00000000200589bb in std::__detail::_Mod_range_hashing::operator() (
    this=0x100006c8950, __num=11876154508389848687, __den=0)
    at /opt/gcc/8.3.0/snos/include/g++/bits/hashtable_policy.h:434
#1  0x00002aaaac492b12 in std::__detail::_Hash_code_base<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> >, std::__detail::_Select1st, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, true>::_M_bucket_index (
    this=0x100006c8950, __c=11876154508389848687, __n=0)
    at /opt/gcc/8.3.0/snos/include/g++/bits/hashtable_policy.h:1392
#2  0x00002aaaac487194 in std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_bucket_index (
    this=0x100006c8950, __k=..., __c=11876154508389848687)
    at /opt/gcc/8.3.0/snos/include/g++/bits/hashtable.h:643
#3  0x00002aaaac498a4d in std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > (this=0x100006c8950, __args#0=..., __args#1=...)
    at /opt/gcc/8.3.0/snos/include/g++/bits/hashtable.h:1673
#4  0x00002aaaac48e3a3 in std::_Hashtable<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > >, std::__detail::_Select1st, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > (this=0x100006c8950, __args#0=..., __args#1=...)
    at /opt/gcc/8.3.0/snos/include/g++/bits/hashtable.h:748
#5  0x00002aaaac48385e in std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int>, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > > >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, unsigned int> > (this=0x100006c8950, 
    __args#0=..., __args#1=...)
    at /opt/gcc/8.3.0/snos/include/g++/bits/unordered_map.h:388
#6  0x00002aaaac46e5a8 in adios2::core::IO::DefineVariable<double> (
    this=0x100006c8900, name=..., shape=..., start=..., count=..., 
    constantDims=false)
    at /global/homes/z/zw241/cworkspace/src/ADIOS2/source/adios2/core/IO.tcc:54
#7  0x00002aaaac8e88f5 in adios2::IO::DefineVariable<double> (
    this=0x100006c8040, name=..., shape=..., start=..., count=..., 
    constantDims=false)
    at /global/homes/z/zw241/cworkspace/src/ADIOS2/bindings/CXX11/adios2/cxx11/IO.tcc:27
#8  0x00000000201129a8 in adiosWrite (fmr=0x100006cdce0, bs=..., 
    inputData=0x2aab74008010, parameters=...)
```

and I don't have any ideas about this error currently, since the code of the data writing is not complicated here.

Thanks a lot for your help !
and this is the adios version that I used
```
commit 9067e90de899a16fb5aa4000d0c75522bdd9c004 (HEAD -> master, origin/master, origin/HEAD)
Merge: e0c14ca8 16eee73d
Author: Greg Eisenhauer <eisen@cc.gatech.edu>
Date:   Wed Feb 26 18:42:31 2020 -0500
```

**Comment by pnorbert** [comment]
I don't understand your multithreaded code but you are not supposed to call DefineVariable multiple times for one variable. The function in IO.tcc should throw an exception if the variable already exist but it doesn't. Did you turn on debug mode for the ADIOS constructor? What kind of error do you see now with the current master, which got rid of the debug mode flag and should throw an error if indeed this is the case here. 

Suggested alternative code is to define the variable once and then call SetSelection() in each thread before calling Put(). We don't officially support multithreaded use of ADIOS but this could work.

**Comment by wangzhezhe** [comment]
Thanks a lot for your help @pnorbert , let me try the current master. BTW,  when you say "turn on debug mode" do you mean setting the related parameter in CMake as "Debug" or there are other parameters to control it.

**Comment by pnorbert** [comment]
No, I meant the boolean flag in the ADIOS() constructor call. What was it
set to? With current master this is irrelevant now because we removed it. I
am just trying to understand what your error was with the older branch.

**Comment by wangzhezhe** [comment]
Do you mean this parameter? `adios2::ADIOS adios(adios2::DebugON);`, let me try current master and use the code that you suggested to see if it works. Thanks! @pnorbert

**Comment by wangzhezhe** [comment]
Hi @pnorbert , if I use the code you mentioned "define the variable once and then call SetSelection() in each thread before calling Put()", how should I set the "start" and the "count" when I define it (different thread is supposed to write different data partition)? Or it doesn't matter since the setSelection will update these two parameters when executing actual data put operation.

**Comment by pnorbert** [comment]
That's what SetSelection() is for.

Watch out for the Put() calls as well. They do nothing in the default deferred mode, just remember the data pointer you pass. Those pointers (and the data behind it) should be valid when EndStep()/PerformPuts() is called. If that is not the case, use Sync-ed Put() calls.

**Comment by germasch** [comment]
@wangzhezhe, it's my understanding that ADIOS2 is not designed to be used from a multithreaded program. It's definitely the case that you will have to call `SetSelection()` and `Put()` from the same thread while locking out other threads from trying to do the same at the same time.

At the very least, I think you have to have a global lock that allows at most one thread inside an ADIOS2 call at a time. In addition, ADIOS2 keeps state (like SetSelection) per Variable, which is a per-process data structure, ie., a second thread calling `SelSelection()` will overwrite the state from the first call. 

I'm not even sure that appropriate usage of a global lock is enough to prevent troubles, it might be that there are races that require it to be always called from the same thread.

**Comment by wangzhezhe** [comment]
Hello, @pnorbert and @germasch , thanks a lot for suggestions. I updated my code by this way:

this is the code that I initialize the adios, and the code is called once per process:

``` 
               //init function
                ...
                adios2::ADIOS adios(adios2::DebugON);
                this->m_io = adios.DeclareIO("gorilla_gs");
                this->m_io.SetEngine("BP4");

                var_u = this->m_io.DefineVariable<double>("U",
                                                               {512, 512, 512},
                                                               {0, 0, 0},
                                                               {512, 512, 512});

                var_step = this->m_io.DefineVariable<int>("step");

                this->m_writer = m_io.Open("gorilla_gs", adios2::Mode::Write);
                ...

        // variables in this struct
        adios2::IO m_io;
        adios2::Engine m_writer;
        adios2::Variable<double> var_u;
        adios2::Variable<int> var_step;
```
This is the code called by every thread:
```
...
    fmr->m_statefulConfig->m_adiosLock.lock();

    adios2::Dims start = {(size_t)bs.m_indexlb[0], (size_t)bs.m_indexlb[1], (size_t)bs.m_indexlb[2]};
    adios2::Dims count = {shapex, shapey, shapez};

    //use set selection to modify the variable selection into the current partition
    fmr->m_statefulConfig->var_u.SetSelection({start, count});

    int step = atoi(parameters[0].c_str());
  
    fmr->m_statefulConfig->m_writer.Put<int>(fmr->m_statefulConfig->var_step, &step, adios2::Mode::Sync);
    fmr->m_statefulConfig->m_writer.Put<double>(fmr->m_statefulConfig->var_u, (double *)inputData, adios2::Mode::Sync);

    fmr->m_statefulConfig->m_adiosLock.unlock();
...
```

I got some errors at step of checking the engine type, it seems that the engine type could not be acquired when execute the data put operation

```
(gdb) #0  0x00002aaab2643f00 in __memcmp_avx2_movbe () from /lib64/libc.so.6
#1  0x00002aaab20704ba in std::char_traits<char>::compare (
    __n=<optimized out>, __s2=0x2aaaac46cb71 "NULL", __s1=<optimized out>)
    at /tmp/peint/build-cray-gcc-20190312.153938.593065186/cray-gcc/BUILD/snos_objdir/x86_64-suse-linux/libstdc++-v3/include/bits/char_traits.h:312
#2  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::compare (this=0x100006ca018, __s=0x2aaaac46cb71 "NULL")
    at /tmp/peint/build-cray-gcc-20190312.153938.593065186/cray-gcc/BUILD/snos_objdir/x86_64-suse-linux/libstdc++-v3/include/bits/basic_string.tcc:1424
#3  0x000000002006d9a8 in std::operator==<char, std::char_traits<char>, std::allocator<char> > (__lhs=..., __rhs=0x2aaaac46cb71 "NULL")
    at /opt/gcc/8.3.0/snos/include/g++/bits/basic_string.h:6084
#4  0x00002aaaac3cf2d4 in adios2::Engine::Put<int> (this=0x100006c8b48, 
    variable=..., data=0x2aaae42a12ec, launch=adios2::Mode::Sync)
    at /global/homes/z/zw241/cworkspace/src/ADIOS2/bindings/CXX11/adios2/cxx11/Engine.tcc:99
```

**Comment by williamfgc** [comment]
@wangzhezhe please find an example for writing from multiple threads to a single bp dataset. 
In the [thread-write example](https://github.com/ornladios/ADIOS2-Examples/blob/master/source/cpp/basics/thread-write.cpp) in the [ADIOS2-Examples repo](https://github.com/ornladios/ADIOS2-Examples). 

It's similar to your case, it's applying lock to the mutex on the I/O parts.
Hope it helps, it's essentially writing a vector<double> with 100 elements, each thread writes a chunk:

```
bpls thread-writes.bp -lavd -n 10
File info:
  of variables:  1
  of attributes: 0
  statistics:    Min / Max 

  double   data  {100} = 0 / 99
    ( 0)    0 1 2 3 4 5 6 7 8 9
    (10)    10 11 12 13 14 15 16 17 18 19
    (20)    20 21 22 23 24 25 26 27 28 29
    (30)    30 31 32 33 34 35 36 37 38 39
    (40)    40 41 42 43 44 45 46 47 48 49
    (50)    50 51 52 53 54 55 56 57 58 59
    (60)    60 61 62 63 64 65 66 67 68 69
    (70)    70 71 72 73 74 75 76 77 78 79
    (80)    80 81 82 83 84 85 86 87 88 89
    (90)    90 91 92 93 94 95 96 97 98 99
```
Each thread writes a block:
```
$ bpls -D thread-writes.bp/
  double   data  {100}
        step 0: 
          block 0: [ 0:11]
          block 1: [12:23]
          block 2: [36:47]
          block 3: [24:35]
          block 4: [48:59]
          block 5: [60:71]
          block 6: [72:83]
          block 7: [84:99]
```

**Comment by germasch** [comment]
Hi @wangzhezhe, I don't see anything wrong with your code, so I'm kinda puzzled. As you already realized, it looks like the engine isn't correctly initialized. (Or maybe it's already destructed?) This doesn't look like it's related to the multi-threading, it seems more like something went wrong when opening the engine.

I think @williamfgc's suggestion to start with a known-working example is a good idea.

**Comment by wangzhezhe** [comment]
Hello, @williamfgc  and @germasch, thanks for the information and let me take some time to dive into the example to see if I missed something in my code.

**Comment by wangzhezhe** [comment]
Hello @williamfgc , there is no error message after I update the code based on your example, it seems that the adios variable is destructed previously and I add a unique pointer this time to fix it. But there is another issue that confuse me a lot,  there is no data written into disk in my case (there is no error message during the program execution):

```
$ ll
total 4
-rw-rw---- 1 zw241 zw241  0 Apr 27 01:52 data.0
-rw-rw---- 1 zw241 zw241  0 Apr 27 01:52 md.0
-rw-rw---- 1 zw241 zw241 64 Apr 27 01:52 md.idx
```
I checked data before data put operation, and there is solid value here. I'm still not sure what are potential reasons that cause this problem (the code you provided works well on my platform).

This is the function that I init the adios objects:

https://github.com/wangzhezhe/Gorilla/blob/master/server/statefulConfig.h#L24

This is the function that the data put is called:

https://github.com/wangzhezhe/Gorilla/blob/master/server/FunctionManager/defaultFunctions/defaultfuncraw.cpp#L220

Thanks a lot for your help!

**Comment by williamfgc** [comment]
@wangzhezhe thanks for the update, I'm glad you made progress. You need to call `Engine::Close()` (outside the threaded region) after you're done with the file. We intentionally don't apply RAII because it's prone to seg faults when mixing the scope of `adios2::Engine` destructors and `MPI_Finalize`. adios2 is primarily a MPI application library and many of the design decisions were driven for getting performance with MPI (including threads, as calling MPI from a thread is not supported in many of the targeted systems). If that changes in the future, we'll need to adapt. Hope it helps.

**Comment by wangzhezhe** [comment]
Got it! Thanks a lot for explanation in detail!@williamfgc  Should I also call `m_engine.Close()` even if I use sync pattern for data put operation? If I call `m_engine.Close`, then if I want to write the data into same bp folder, should I open that engine again?

**Comment by williamfgc** [comment]
@wangzhezhe thanks for the question. Short answer: yes. Long answer: `Put` manages your memory interactions with adios2, doesn't mean it's doing real disk I/O as traditional `write`. `Put` puts (pun not intended) things on a buffer (another reason for not supporting thread-safety: single buffering has little opportunity for threading due to is inherently serial nature, in fact we call the class `BP4Serializer`) as we try to minimize contact with the file system. `Close` is essentially the final point of contact with each Engine in your app. 

Since you already digested a lot of the `adios2` concepts, please take a look at the [docs for Put  memory contracts](https://adios2.readthedocs.io/en/latest/components/components.html#put-modes-and-memory-contracts). Let us know if you have any questions. Hope it helps.

**Comment by williamfgc** [comment]
> If I call m_engine.Close, then if I want to write the data into same bp folder, should I open that engine again?

Yes, but `Open` it with `adios2::Mode::Append`.  A good practice is to keep your file Open for as much as you can, ideally having only one `Open/Close` pair per engine. We use BeginStep/EndStep for streaming style, in your threaded case that would require a Barrier and calling those from a single thread id.

**Comment by wangzhezhe** [comment]
Thanks so much for the information! @williamfgc
